Improve feedback for correctness tests

rabidbit opened the pull request:
Work-in-progress. Needs tests. Looking for initial feedback.

seanlip commented on client/question/components/LearnerViewDirective.js:
```diff
@@ -59,6 +59,10 @@ tie.directive('learnerView', [function() {
                         </syntax-error-snippet>
                       </span>
+                      <span ng-if="paragraph.isOutputParagraph()">
+                        <output-snippet content="paragraph.getContent()">
```
I don't see the file defining an OutputSnippet directive. Did you forget to "git add"?

(As a general rule, I suggest checking the ["Files Changed" tab](https://github.com/google/tie/pull/438/files) after pushing to GitHub, in order to ensure that the diff is exactly what you want it to be.)

rabidbit replied:
Ah... I forgot to git add the new file. I'll add it.

rabidbit replied:
I think I addressed this?

seanlip commented on client/question/domain/FeedbackObjectFactory.js:
```diff
@@ -139,6 +139,17 @@ tie.factory('FeedbackObjectFactory', [
+     * Appends a FeedbackParagraph of type output to the _paragraphs
+     * Array
+     *
+     * @param {string} text String of code to be inserted as code output
```
Perhaps `code` might be a better variable name? Or even `outputString`? Unclear to me, from reading the docstring, whether this is student's code or the output printout from Python -- I'd assume the latter, but that isn't really code.

rabidbit replied:
outputString SGTM. Will change it to that.

seanlip commented on client/question/domain/FeedbackObjectFactory.js:
```diff
@@ -139,6 +139,17 @@ tie.factory('FeedbackObjectFactory', [
 
     /**
+     * Appends a FeedbackParagraph of type output to the _paragraphs
+     * Array
```
Minor nit, here and below: end these docstrings and arg descriptions with periods. (I think this was missed in some cases previously.)

rabidbit replied:
Yeah, it was copy/paste from existing code. Will add periods to the  existing code.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -24,18 +24,26 @@ tie.factory('FeedbackGeneratorService', [
+
+    /**
+     * Object used to keep track of how many corrective feedback (values)
+     * were generated for a given test suite (keys).
```
This is a bit vague -- "corrective feedback" could refer to buggy output feedback etc. We should pick a term for the type of feedback you're implementing and stick to it in all docstrings/varnames. (It's fine if the docstring ends up being a bit stilted, as long as it's unambiguous what it's referring to.)

rabidbit replied:
I meant "correctness" (typo). Is "correctness" OK? I don't have strong feelings about the name.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -24,18 +24,26 @@ tie.factory('FeedbackGeneratorService', [
+     *
+     * @type {Object.<string, number>}
+     */
+    var attemptCounter = {};
```
Ditto; the name needs to be more specific. Perhaps suiteIdToNumAttempts or similar. That said -- this is not the same as the total number of attempts made by the learner, right? It only counts those attempts which "trickle down" to the correctness-test level -- and that needs to be made clearer in the name + docstring.

Thinking about it more, I'd recommend a change in framing here: instead of mapping suite ID to number of attempts, instead map suite ID to one of three named string constants denoting the stages of feedback (CORRECTNESS_TEST_INPUT_GIVEN, CORRECTNESS_TEST_EXPECTED_OUTPUT_GIVEN, CORRECTNESS_TEST_OBSERVED_OUTPUT_GIVEN). This is so we just store the minimal "status" information necessary to give the appropriate feedback and determine the next status to use.

rabidbit replied:
How about correctnessFeedbackCounterBySuiteId?

I'm not sure I follow your suggestion. This appears to be more complex and granular than for the purpose I was using this.

rabidbit replied:
Will update per discussion.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -24,18 +24,26 @@ tie.factory('FeedbackGeneratorService', [
+     *
+     * @type {Object.<string, number>}
+     */
+    var attemptCounter = {};
```
Two general design questions:

- If the user leaves the page and comes back, will this status be preserved? I'm guessing not, and that may be fine for now, but I just wanted to point that out.
- If the user gets an error for test 1 in suite A, and proceeds to the "expected output given" step, then fixes that error but their code is wrong for test 3 in suite A (despite passing test 1 in suite A), what should happen? This counter collapses all status data into a per-test-suite dict and it's not clear to me whether that correctly handles the case I've just described.

rabidbit replied:
- That's true. I agree it's probably fine for this PR though.
- Argh.. There was a regression check that was in there before; I must have accidentally deleted it when removing the hint related code. It checked whether the user failed on an earlier test case they had previously passed. I'll write it in again. When you say "test", are you referring to testCases as a whole or the individual input/allowedOutput pairs within a testCase?

rabidbit replied:
Will update per discussion.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +61,57 @@ tie.factory('FeedbackGeneratorService', [
+    var lastCorrectnessFeedback = {};
+
+    /**
+     * Index of last test suite for which corrective feedback was given. Used
```
Systematize terminology -- call it either "correctness feedback" or "corrective feedback", but not both in different places. I think I lean towards the former, since that's what we've historically been using.

rabidbit replied:
Yep, another typo (I think I was sleepy when writing some of this :-P). I meant "correctness". Will do.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +61,57 @@ tie.factory('FeedbackGeneratorService', [
     /**
+     * Object containing variations on feedback text for different scenarios.
+     *
+     * $type {Object.<string, Array.<string>>}
```
Follow existing docstring conventions please -- `@type` not `$type`. Ditto below.

rabidbit replied:
Ah, that was typo.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +61,57 @@ tie.factory('FeedbackGeneratorService', [
+     *
+     * $type {Object.<string, Array.<string>>}
+     */
+    var correctnessFeedbackText = {
```
Pull out into constant in questions.js, rename to CORRECTNESS_FEEDBACK_TEXT. Upper-caps the interior keys as well (these could be equal to the status names I suggested above).

Makes it easier to change just the text in the future, if needed.

rabidbit replied:
Yeah, this was originally a constant, but I went back and forth and eventually made it a regular variable, thinking it could be updated. OK, I'll change it back to constant and move it to questions.s.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +121,39 @@ tie.factory('FeedbackGeneratorService', [
+     * of a given feedback type.
+     *
+     * @param {Object.<string, Array.<string>>} correctnessFeedbackText
+     * @param {string} type Correctnness feedback type.
```
Typo in "Correctness".

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +121,39 @@ tie.factory('FeedbackGeneratorService', [
+     * @returns {string}
+     * @private
+     */
+    var _selectCorrectnessFeedback = function(correctnessFeedbackText, type) {
```
Consider _getCorrectnessFeedbackString; the internal details that this is selection from a random list don't matter to the caller.

Why is correctnessFeedbackText being passed in as a variable, btw? It's global in this scope and I don't think it needs to be passed in explicitly.

rabidbit replied:
Changed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +121,39 @@ tie.factory('FeedbackGeneratorService', [
+     * @private
+     */
+    var _selectCorrectnessFeedback = function(correctnessFeedbackText, type) {
+      var defaultFeedbackIndex = lastCorrectnessFeedback[type];
```
As a general rule, if the only thing we're referring to in lastCorrectnessFeedback is the type, then we should just store the type. (That said, that variable might not be needed any more if we're keeping track of status as described above.)

rabidbit replied:
This is used to make sure we don't present the same feedback message twice. There is an index tied to lastCorrectnessFeedback; lastCorrectnessFeedback[type] stores the index of the feedback last given for the specified type. Does that make sense?

seanlip replied:
In that case, how about just storing a lastCorrectnessFeedbackString instead? If I'm understanding correctly, I don't think this needs to be segmented by type...

rabidbit replied:
I might not have been precise enough, I am not showing the same feedback two times in a row for a given feedback type. E.g., 
feedback-type-a-1
feedback-type-b-1
feedback-type-a-2 (feedback-type-a-1 not allowed since it was last used)

Does that make sense?

seanlip replied:
Not quite, sorry. I assume a and b refer to distinct states/statuses, and b comes later in the sequence than a. If that's the case, what I don't understand is why we'd go back to feedback of type a? I thought in the state diagram we discussed, the user can only move forward through it.

rabidbit replied:
I wanted to implement variation across correctness tests as well.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -268,28 +360,45 @@ tie.factory('FeedbackGeneratorService', [
     var _getCorrectnessTestFeedback = function(
-      outputFunctionName, testCase, observedOutput) {
+      outputFunctionName, testCase, observedOutput, testSuiteId,
+      testSuiteIndex) {
```
Please update the docstring when adding new args.

rabidbit replied:
Done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -268,28 +360,45 @@ tie.factory('FeedbackGeneratorService', [
+      if (testSuiteIndex == lastTestSuiteIndex) {
+        // Same test suite; attempt number > 1
+        attemptCounter[testSuiteId] += 1;
+        if (attemptCounter[testSuiteId] == 2 && testSuiteId != 'SAMPLE_INPUT') {
```
Don't special case on SAMPLE_INPUT; this should be generic. If buggy/suite-level feedback exists for SAMPLE_INPUT, it will show up prior to correctness feedback being invoked anyway. So there's no need to special-case it here.

rabidbit replied:
Per discussion, make this a global constant.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -268,28 +360,45 @@ tie.factory('FeedbackGeneratorService', [
+        attemptCounter[testSuiteId] += 1;
+        if (attemptCounter[testSuiteId] == 2 && testSuiteId != 'SAMPLE_INPUT') {
+          // Display expected output
+          feedback.appendTextParagraph(
```
Minor style nit: when a list of multiple args spans more than one line, break after the first '('. So, here, you could do:

              feedback.appendTextParagraph(_selectCorrectnessFeedback(
                correctnessFeedbackText, 'expected_output'));

rabidbit replied:
done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -268,28 +360,45 @@ tie.factory('FeedbackGeneratorService', [
-      feedback.appendCodeParagraph(
-        _jsToHumanReadable(allowedOutputExample));
-      feedback.appendTextParagraph('Could you fix this?');
+      // Allow user to display output
```
Ah, I thought we were only going to do this after they've passed a "only sample input" and "only sample input + expected output" stage? Otherwise I suspect it'll be too much of a crutch....

EDIT: sorry, misread, I missed that the two previous clauses actually returned stuff. I think if you get rid of the SAMPLE_INPUT special-casing then you can put the stuff below in an else branch to more clearly indicate it's an alternative to the previous two scenarios, rather than an add-on.

rabidbit commented:
Updated to address most of the comments. Main thing left is to perhaps merge OutputSnippetDirective.js with SyntaxErrorSnippetDirective.js.

seanlip commented on client/question/domain/FeedbackObjectFactory.js:
```diff
@@ -129,13 +129,25 @@ tie.factory('FeedbackObjectFactory', [
+     * Array.
      *
-     * @param {string} text String describing the syntax error
+     * @param {string} outputString String describing the syntax error.
```
I think you meant to rename the method below, not this one...

rabidbit replied:
Ah. Sorry. Fixed.

seanlip commented on client/question/domain/FeedbackParagraphObjectFactory.js:
```diff
@@ -19,8 +19,9 @@
 
 tie.factory('FeedbackParagraphObjectFactory', [
   'PARAGRAPH_TYPE_TEXT', 'PARAGRAPH_TYPE_CODE', 'PARAGRAPH_TYPE_SYNTAX_ERROR',
+  'PARAGRAPH_TYPE_OUTPUT',
```
Have line breaks here be the same as the ones in the function signature below -- this makes it easier to visually check that the dependencies line up.

rabidbit replied:
done

seanlip commented on client/question/domain/FeedbackParagraphObjectFactory.js:
```diff
@@ -82,6 +84,15 @@ tie.factory('FeedbackParagraphObjectFactory', [
     };
 
     /**
+     * Checks if the Paragraph is a paragraph consisting of code output
```
Nit: add period at end.

rabidbit replied:
done

seanlip commented on client/question/question.js:
```diff
@@ -20,6 +20,71 @@ window.tie = angular.module('tie',
+ *
+ * @type {string}
+ */
+tie.constant('SAMPLE_INPUT', 'SAMPLE_INPUT'); 
```
For the name of the constant, include what type of thing this is (in this case, it's a test suite ID, so maybe you could call it e.g. TEST_SUITE_ID_SAMPLE_INPUT).

rabidbit replied:
done

seanlip commented on client/question/question.js:
```diff
@@ -20,6 +20,71 @@ window.tie = angular.module('tie',
+ *
+ * @type {string}
+ */
+tie.constant('CORRECTNESS_TEST_INPUT_DISPLAYED',
```
Ditto; start with the type of what this is. Maybe CORRECTNESS_STATE_KEY_*.

rabidbit replied:
I thought I was following the naming convention you suggested for string constants denoting the stages of feedback? I used 'DISPLAYED' instead of 'GIVEN'.

seanlip replied:
Ah yep, this one for example should be CORRECTNESS_STATE_KEY_INPUT_DISPLAYED (that's what I meant by the previous notation).

rabidbit replied:
Just to make sure we talking about the same thing, you suggested in a prior comment to use:
"CORRECTNESS_TEST_INPUT_GIVEN, CORRECTNESS_TEST_EXPECTED_OUTPUT_GIVEN, CORRECTNESS_TEST_OBSERVED_OUTPUT_GIVEN". I changed GIVEN to DISPLAYED but used "CORRECTNESS_TEST_*". Are now you saying use "CORRECTNESS_STATE_KEY_*" instead or is this something different? If we are talking about the same thing, would "CORRECTNESS_STATE_*" work or do you think we really should have "_KEY" as well? Though I don't have strong feelings about it.

rabidbit replied:
Not sure why "Are now you saying use" is italicized; not trying to emphasize it. Can't seem to change it back to normal text.

seanlip replied:
It's a markdown thing; enclosing anything between asterisks italicizes it. A fix would be to use backticks (`) to surround parts that should be displayed in code form.

I also got the name wrong earlier, sorry. I think CORRECTNESS_STATE or CORRECTNESS_STATE_KEY would be fine. Happy with the former.

rabidbit replied:
Changed to 'CORRECTNESS_STATE_*'

seanlip commented on client/question/question.js:
```diff
@@ -20,6 +20,71 @@ window.tie = angular.module('tie',
+ * @type {Object.<string, Array.<string>>}
+ */
+tie.constant('CORRECTNESS_FEEDBACK_TEXT', {
+  'INPUT_TO_TRY': [
```
These keys could match the values in the constants above?

E.g. above, define: tie.constant('CORRECTNESS_STATE_KEY_INPUT_DISPLAYED', 'INPUT_DISPLAYED');

and here:

```
tie.constant('CORRECTNESS_FEEDBACK_TEXT', {
  INPUT_DISPLAYED: [...]
});
```

seanlip commented on client/question/question.js:
```diff
@@ -215,6 +283,14 @@ tie.constant('PARAGRAPH_TYPE_CODE', 'code');
 tie.constant('PARAGRAPH_TYPE_SYNTAX_ERROR', 'error');
 
 /**
+ * FeedbackParagraph type that will render text to look like code.
```
This sounds like PARAGRAPH_TYPE_CODE to me, FYI.

rabidbit replied:
They are similar, but code is used as formatting text for things like buggy output. This type indicates the display of the content can be toggled (more than just formatting text). Let me know if you want to handle this differently.

seanlip replied:
I suppose if we merge syntax-error and output paragraphs, this will give us a feedback paragraph type whose duty is to render text with a "display more" button?

rabidbit replied:
If you're talking about syntax and output types, yes, that sounds like a possibility. I thought you were talking about syntax and code types, which I think are more different in terms behavior.

seanlip replied:
Ah yes. I just meant the docstrings for both were the same and that implies a false equivalence (as discussed offline in chat).

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -24,18 +24,32 @@ tie.factory('FeedbackGeneratorService', [
+    CORRECTNESS_TEST_OBSERVED_OUTPUT_DISPLAYED, CORRECTNESS_FEEDBACK_TEXT) {
+
+    /**
+     * Object used to keep track of how many correctness feedback (values)
```
I thought we decided that this was going to have values that were strings representing states, as opposed to numeric counts?

rabidbit replied:
We are. I just forgot to update the variable description. Fixed.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +93,40 @@ tie.factory('FeedbackGeneratorService', [
+    var _getRandomInt = function(min, max) {
+      min = Math.ceil(min);
+      max = Math.floor(max);
+      return Math.floor(Math.random() * (max - min)) + min
```
Missing semicolon at EOL.

rabidbit replied:
Thanks for the catch! Fixed.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +93,40 @@ tie.factory('FeedbackGeneratorService', [
+     * @returns {string}
+     * @private
+     */
+    var _getCorrectnessFeedbackString = function(correctnessFeedbackText,
```
Nit: Insert line break after '('.

Also, I still don't understand why correctnessFeedbackText is being passed in. Can't this method refer directly to CORRECTNESS_FEEDBACK_TEXT?

rabidbit replied:
Changed line break.

Re: correctnessFeedbackText vs CORRECTNESS_FEEDBACK_TEXT, I just forgot to change the variable name to the new one. Fixed.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +93,40 @@ tie.factory('FeedbackGeneratorService', [
+      while (defaultFeedbackIndex == lastCorrectnessFeedback[type]) {
+        var defaultFeedbackIndex = _getRandomInt(0,
+          (correctnessFeedbackText[type].length));
+        }
```
Deindent by 2 to match opening brace.

rabidbit replied:
done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +93,40 @@ tie.factory('FeedbackGeneratorService', [
+        type) {
+      var defaultFeedbackIndex = lastCorrectnessFeedback[type];
+      while (defaultFeedbackIndex == lastCorrectnessFeedback[type]) {
+        var defaultFeedbackIndex = _getRandomInt(0,
```
Ditto, break after '('. No need for inner parens on the next line, too.

rabidbit replied:
done.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
-      outputFunctionName, testCase, observedOutput) {
+        outputFunctionName, testCase, testCaseIndex, observedOutput,
+        testSuiteId) {
+      var testCaseStateKey = testSuiteId + '-' + testCaseIndex;
```
This key just encodes a description of the test case, not its state, so you can call it testCaseKey.

rabidbit replied:
done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
-            _jsToHumanReadable(observedOutput)));
-        feedback.appendTextParagraph(
-          'However, the expected output of ' + outputFunctionName + ' is:');
+      if (testSuiteId == lastTestSuiteId) {
```
I keep reading "lastTestSuiteId" as "the last test suite defined in the question". Could we say "previousTestSuiteId" instead?

rabidbit replied:
done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
+      if (testSuiteId == lastTestSuiteId) {
+        if (correctnessTestStates[testCaseStateKey] ==
+            CORRECTNESS_TEST_INPUT_DISPLAYED &&
+            testSuiteId != SAMPLE_INPUT) {
```
Having SAMPLE_INPUT in here is a bit convoluted. I recommend the following:

- At the outset, make sure the state is correct:
  - Check whether the state doesn't exist, if so, initialize it.
  - Check whether the suite ID is sample input; if it is then set the current state to be OBSERVED_OUTPUT_DISPLAYED.

Now, you can just act based only on the state, and do `if (state == SUCCESS) {...} else if (state == OUTPUT_ENABLED) {...} else if (state == EXPECTED_OUTPUT_DISPLAYED) {...} else if (state == INPUT_DISPLAYED) {...} else {throw Error('...')}`. This helps split up the logic and makes it easier to maintain this in the future.

rabidbit replied:
Let me think about this. I'm not sure setting OBSERVED_OUTPUT_DISPLAYED on SAMPLE_INPUT would work, given it would bypass the hinting phase. I'll think about how to make this simpler and more readable.

rabidbit replied:
Updated in latest commit.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
+              'INPUT_TO_TRY'));
+            feedback.appendCodeParagraph(
+              'Input: ' + _jsToHumanReadable(testCase.getInput()));
+            correctnessTestStates[testSuiteId] = 1;
```
This looks incorrect...

rabidbit replied:
Forgot to delete that line. Thanks for the catch.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
+        // New / next test suite
+        lastTestSuiteId = testSuiteId;
+        if (testSuiteId != SAMPLE_INPUT) {
+          if (!(testCaseStateKey in correctnessTestStates)) {
```
if (!correctnessTestStates.hasOwnProperty(testCaseStateKey)) {

rabidbit replied:
done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,33 +328,70 @@ tie.factory('FeedbackGeneratorService', [
+            return feedback;
+          } else {
+            // Catch regressions
+            feedback.appendTextParagraph('It looks like there was a ' +
```
Nit: break after '('

rabidbit replied:
done

rabidbit commented:
Addressed most comments. Outstanding: (a) see about refactoring the output logic to make it more readable, (b) investigate why feedback goes back and forth between suite level feedback and correctness feedback, (c) add tests

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +93,40 @@ tie.factory('FeedbackGeneratorService', [
+     * @private
+     */
+    var _getCorrectnessFeedbackString = function(
+        CORRECTNESS_FEEDBACK_TEXT, type) {
```
This wasn't what I meant when I wrote the previous comment, I think. What I was trying to say was that you do not need to pass in CORRECTNESS_FEEDBACK_TEXT as an argument at all.

rabidbit replied:
ah, got it

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
+      if (testSuiteId != previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
```
Just to check, is the logic here correct? For example, if the user has passed suite 1 and 2 without getting any feedback, and subsequently regresses, would there even be an entry for those test cases in correctnessTestStates?

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
+        }
+        correctnessTestStates[testCaseKey] = CORRECTNESS_STATE_STARTING;
+      }
+      // Check if sample input test suite (input and expected already displayed)
```
This is quite cryptic; maybe be more descriptive. E.g.:

```
// If the suite ID corresponds to the sample input, the question will already have
// displayed the input and expected output, so we advance the correctness state
// to "expected output displayed".
```

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
+          return feedback;
+          break;
+        default:
+          // Allow user to display code output
```
... to display the output of their code.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
+        case CORRECTNESS_STATE_STARTING:
+          // Display input user should try
+          feedback.appendTextParagraph(
+            _getCorrectnessFeedbackString('INPUT_TO_TRY'));
```
It would be nice if the string literals here could be replaced with the already-named CORRECTNESS_STATE_* constants (and the object constant modified accordingly); we should also add a Karma test to ensure that the keys of CORRECTNESS_FEEDBACK_TEXT match the existing constants.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -263,34 +328,75 @@ tie.factory('FeedbackGeneratorService', [
-      outputFunctionName, testCase, observedOutput) {
+        outputFunctionName, testCase, testCaseIndex, observedOutput,
+        testSuiteId) {
+      var testCaseKey = testSuiteId + '-' + testCaseIndex;
```
Pull this logic out into a separate helper function: getTestCaseKey(testSuiteId, testCaseIndex)

seanlip commented:
(Also, please note that e2e tests are failing.)

rabidbit commented:
@seanlip PTAL.

seanlip commented:
Hi @rabidbit -- I'll take a look at this within the next hour, but FYI you might want to take a look at the codecov reports in the meantime. It looks like test coverage has dropped quite a bit with this change, so it might be worth taking a look at the red bits and ensuring those are covered by tests. Thanks!

seanlip commented:
Also, @rabidbit -- please could you address my most recent round of comments? https://github.com/google/tie/pull/438/#pullrequestreview-73621377

seanlip commented on client/question/question.js:
```diff
@@ -20,6 +20,79 @@ window.tie = angular.module('tie',
+ */
+tie.constant('CORRECTNESS_FEEDBACK_TEXT', {
+  INPUT_TO_TRY: [
+    'Would your code work for the following input?',
```
@eyurko Would you mind doing a quick review of the text here?

seanlip commented:
(Adding @eyurko for review of text in questions.js)

seanlip commented on client/question/question.js:
```diff
@@ -20,6 +20,79 @@ window.tie = angular.module('tie',
+ * @type {Object.<string, Array.<string>>}
+ */
+tie.constant('CORRECTNESS_FEEDBACK_TEXT', {
+  INPUT_TO_TRY: [
```
Re-flagging a previous comment asking if we could have these names match the correctness state constants above (and adding a Karma test to ensure that the keys of CORRECTNESS_FEEDBACK_TEST are exactly those constants).

rabidbit replied:
Think this was addressed.

seanlip commented on client/question/question.js:
```diff
@@ -215,6 +291,14 @@ tie.constant('PARAGRAPH_TYPE_CODE', 'code');
 tie.constant('PARAGRAPH_TYPE_SYNTAX_ERROR', 'error');
 
 /**
+ * FeedbackParagraph type that will render text to look like code.
```
Are we fixing this? @rabidbit

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -24,18 +24,34 @@ tie.factory('FeedbackGeneratorService', [
+
+    /**
+     * Object used to keep track of which state we are in for correctness
+     * feedback.
```
This docstring could use more detail; what are the keys and values? (In particular I think info about the keys is missing.)

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +69,23 @@ tie.factory('FeedbackGeneratorService', [
 
     /**
+     * Object to hold index of correctness feedback given (values), per
+     * test suite (keys). Used when avoiding presenting the same feedback text
```
when avoiding --> in order to avoid

rabidbit replied:
Changed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +69,23 @@ tie.factory('FeedbackGeneratorService', [
 
     /**
+     * Object to hold index of correctness feedback given (values), per
+     * test suite (keys). Used when avoiding presenting the same feedback text
```
A test suite is an object, so it can't be a key. Do you mean test suite ID, test case key, or something else?

What does "index of correctness feedback given" mean?

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +95,38 @@ tie.factory('FeedbackGeneratorService', [
+     * Randomly selects feedback string from an object containing variations
+     * of a given feedback type.
+     *
+     * @param {string} type Correctness feedback type.
```
If this is functionally identical to state then let's just use the terminology "state" throughout (and call this arg correctnessFeedbackState or similar).

rabidbit replied:
I don't think they are functionally identical. I also feel using "state" to label feedback text type is actually more confusing.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +69,23 @@ tie.factory('FeedbackGeneratorService', [
     /**
+     * Object to hold index of correctness feedback given (values), per
+     * test suite (keys). Used when avoiding presenting the same feedback text
+     * consecutively.
```
If the only thing we care about here is not showing the same feedback consecutively, is there a reason to not just use a string instead? If a string would suffice, we don't even need to store this here -- we could just extract the previous feedback given from the TranscriptService.

An alternative (if we really want to store the index) is to store two vars: the previous type and the previous index. Only if both match, do we regenerate.

rabidbit replied:
The intention is to not show the same feedback twice in a row, even if the user got another piece of feedback in between. I.e., not only is the following avoided:
1. feedback 1 of type a
2. feedback 1 of type a
The following is also avoided:
1. feedback 1 of type a
2. feedback 1 of type b
3. feedback 1 of type a

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +95,38 @@ tie.factory('FeedbackGeneratorService', [
+     */
+    var _getCorrectnessFeedbackString = function(type) {
+      var correctnessFeedbackIndex = previousCorrectnessFeedback[type];
+      while (correctnessFeedbackIndex === previousCorrectnessFeedback[type]) {
```
This is ... OK, I guess. It would be slightly better to generate a random int from [0, arr.length - 1] and adjust the index accordingly.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -204,7 +269,7 @@ tie.factory('FeedbackGeneratorService', [
 
         if (previousFeedback.getFeedbackCategory() ===
-            currentFeedbackCategory) {
+          currentFeedbackCategory) {
```
I think the previous indentation is preferable, actually. It makes it easier to tell that the next line is the start of the child scope, and it left-aligns with the '(' on the previous line (which is generally what to do when the contents of a parentheses span multiple lines).

rabidbit replied:
I think I changed this to address a linter error on indentation.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
     };
 
     /**
+     * Concats two strings to create testCaseKey
```
Nits:
- Concats --> Concatenates
- Add period at end.
- The main purpose of this function (from an outsider's perspective) is to create a key that uniquely identifies a test case. So I wouldn't worry about the implementation details; a better docstring would be "Generates a unique test case key for the given (test suite, test case) pair."

rabidbit replied:
Changed to suggested text.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
      * @param {CorrectnessTest} TestCase
+     * @param {number} testCaseIndex
      * @param {*} observedOutput Actual output for running user's code.
+     * @param {string} testSuiteId
```
Optional: I'd like to suggest putting this just before testCaseIndex in the argument ordering, since logically the suite comes before the test case.

rabidbit replied:
Changed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
+        if (correctnessTestStates.hasOwnProperty(testCaseKey)) {
+          feedback.appendTextParagraph(
+            'It looks like there was a regression in your code. Your code ' +
+            'worked for the following, but it now fails:');
```
Perhaps "worked" --> "used to work"

rabidbit replied:
Changed as requested.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
+      // Check if new / next test suite
+      if (testSuiteId !== previousTestSuiteId) {
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions
```
Copying comment from previous review:

> Just to check, is the logic here correct? For example, if the user has passed suite 1 and 2 without getting any feedback, and subsequently regresses, would there even be an entry for those test cases in correctnessTestStates?

rabidbit replied:
It should not. IIUC, if they get no feedback at all, then they would also get no correctness feedback. In that case, there should be no entry in correctnessTestStates. If they fail a test later, it would be as if it was the first time (not a regression); at that point, an entry is made in correctnessTestStates. Since testCaseKey includes both the suite ID and test case index, it should hold across different test suites.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
+          CORRECTNESS_STATE_EXPECTED_OUTPUT_DISPLAYED;
+      }
+      switch (correctnessTestStates[testCaseKey]) {
+      case CORRECTNESS_STATE_STARTING:
```
Indent this and the various lines below by 2 (until the closing "}" character), to indicate that this is an inner scope.

rabidbit replied:
That's what I had originally, but I think I changed it to address linter errors. I'll try to change it back to see if the errors still show.

rabidbit replied:
Lint errors prevent this.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
+      }
+      switch (correctnessTestStates[testCaseKey]) {
+      case CORRECTNESS_STATE_STARTING:
+        // Display input user should try
```
Display an input that the learner should use to manually walk through their code.

Also, add periods at ends of comments, here and below.

rabidbit replied:
Changed as requested.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +322,87 @@ tie.factory('FeedbackGeneratorService', [
+        feedback.appendTextParagraph(
+          _getCorrectnessFeedbackString('OUTPUT_ENABLED'));
+        feedback.appendOutputParagraph(
+          'Input: ' + _jsToHumanReadable(testCase.getInput()) +
```
Just to check, did you also mean to show the input and expected output when the display is toggled? I thought at this point the input / expected output would always be shown.

rabidbit replied:
This is to provide context for the output displayed. When we implement hints, hints can be displayed before the option to display output; so, at that point, we want to provide the input and expected output as context.

seanlip commented on client/question/components/OutputSnippetDirective.js:
```diff
@@ -0,0 +1,108 @@
+        $scope.snippetLines = [];
+
+        /**
+         * Represents whether a syntax error is being shown or not.
```
This docstring is incorrect. Please verify all docstrings in the file.

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -324,7 +436,7 @@ tie.factory('FeedbackGeneratorService', [
      */
     var _getRuntimeErrorFeedback = function(
-        codeEvalResult, rawCodeLineIndexes) {
+      codeEvalResult, rawCodeLineIndexes) {
```
It's OK to leave this indented by two additional spaces (see reasoning in an earlier comment).

rabidbit replied:
As mentioned, indentation were changed to address linter errors. I will try to revert to original spacing (perhaps after addressing all the other comments and before final approval/merge).

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -336,7 +448,7 @@ tie.factory('FeedbackGeneratorService', [
             Number(humanReadableLineNumber) - 1);
           if (preprocessedCodeLineIndex < 0 ||
-              preprocessedCodeLineIndex >= rawCodeLineIndexes.length) {
+            preprocessedCodeLineIndex >= rawCodeLineIndexes.length) {
```
Ditto.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -519,11 +631,12 @@ tie.factory('FeedbackGeneratorService', [
             for (var k = 0; k < testCases.length; k++) {
               var testCase = testCases[k];
+              var testCaseIndex = k;
+              var testSuiteId = testSuites[j]._id;
```
Never reference a private member like this. Use getId() instead.

rabidbit replied:
Fixed.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -519,11 +631,12 @@ tie.factory('FeedbackGeneratorService', [
             var testCases = testSuites[j].getTestCases();
             for (var k = 0; k < testCases.length; k++) {
               var testCase = testCases[k];
+              var testCaseIndex = k;
```
You could just replace k with testCaseIndex in the for loop definition in line 632 -- in general, we try to avoid having two different names/variables for the same thing.

rabidbit replied:
Changed.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -171,61 +167,98 @@ describe('FeedbackGeneratorService', function() {
+      testCase: {
+        input: 'Hello, John',
+        allowedOutputs: ['olleH, nhoJ']
+      }};
```
Optional: consider dropping the last `};` to the next line so that it matches the indentation level of its corresponding start brace. Ditto below.

rabidbit replied:
Changed.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -171,61 +167,98 @@ describe('FeedbackGeneratorService', function() {
+      function() {
+        var correctnessFeedbackParagraphs =
+          FeedbackGeneratorService._getCorrectnessTestFeedback(
+          sampleInputTestCase, 0, 'incorrect answer',
```
Indent this and the next line by 2.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -171,61 +167,98 @@ describe('FeedbackGeneratorService', function() {
+          FeedbackGeneratorService._getCorrectnessTestFeedback(
+          sampleInputTestCase, 0, 'incorrect answer',
+          sampleInputTestSuite.id).getParagraphs();
+        var expectedCorrectnessFeedbackText =
```
I'm afraid this isn't the right way to do this -- the expected/actual values should not need additional massaging, and setting "actual = expected" at the outset is a bit of a red flag. Here is how I would suggest writing the next 28 lines or so:

```
        var firstFeedbackParagraphText = correctnessFeedbackParagraphs[0].getContent();
        var possibleFeedbackTexts = CORRECTNESS_FEEDBACK_TEXT.OUTPUT_ENABLED;
        expect(possibleFeedbackTexts.includes(firstFeedbackParagraphText)).toEqual(true);

        expect(correctnessFeedbackParagraphs[0].isTextParagraph()).toBe(true);
        expect(correctnessFeedbackParagraphs[1].isOutputParagraph()).toBe(true);
        expect(correctnessFeedbackParagraphs.length).toEqual(2);
```

Similar feedback applies for other tests. Basically, don't declare proxy values and try to set them with complicated logic; instead, try to keep the tests as clean and straightforward as possible.

rabidbit replied:
The idea was to make it clear what the problem was. I would see vague failure messages saying the expected value of 5 was 2 or the expected value of true was false, making it very hard to look at the failure message and understand what exactly went wrong. That said, I'll change it as you suggested to make it consistent.

seanlip commented on client/question/services/SolutionHandlerServiceSpec.js:
```diff
@@ -454,8 +456,18 @@ describe('SolutionHandlerService', function() {
           ).then(function(feedback) {
-            expect(feedback.getParagraphs()[0].getContent()).toBe(
-              'Your code produced the following result:');
+            var expectedFeedbackTextResult = 'valid feeback text';
```
Note that 'feedback' in the string constant is misspelled.

rabidbit commented on client/question/question.js:
```diff
@@ -20,6 +20,79 @@ window.tie = angular.module('tie',
+  'CORRECTNESS_STATE_INPUT_DISPLAYED');
+
+/**
+ * Correctness state where the expeced output has been displayed to the user.
```
expeced misspelled

rabidbit replied:
Fixed

rabidbit commented:
@seanlip Yeah, I see errors when I indent as you requested (and as I originally had them). E.g.

```
/home/travis/build/google/tie/client/question/services/FeedbackGeneratorService.js
  400:9   error  Expected indentation of 6 spaces but found 8   indent
  403:11  error  Expected indentation of 8 spaces but found 10  indent
  405:11  error  Expected indentation of 8 spaces but found 10  indent
  407:11  error  Expected indentation of 8 spaces but found 10  indent
  409:11  error  Expected indentation of 8 spaces but found 10  indent
  410:9   error  Expected indentation of 6 spaces but found 8   indent
  412:11  error  Expected indentation of 8 spaces but found 10  indent
  414:11  error  Expected indentation of 8 spaces but found 10  indent
  418:11  error  Expected indentation of 8 spaces but found 10  indent
  420:11  error  Expected indentation of 8 spaces but found 10  indent
  421:9   error  Expected indentation of 6 spaces but found 8   indent
  423:11  error  Expected indentation of 8 spaces but found 10  indent
  425:11  error  Expected indentation of 8 spaces but found 10  indent
  429:11  error  Expected indentation of 8 spaces but found 10  indent
  431:11  error  Expected indentation of 8 spaces but found 10  indent
```

seanlip commented:
Ah ok, thanks. That's weird, I don't know why they haven't triggered historically. But yeah, in that case, let's do what's needed to make them go away.

Thanks for checking!

rabidbit commented:
PTAL

eyurko commented on client/question/question.js:
```diff
@@ -20,6 +20,101 @@ window.tie = angular.module('tie',
+  INPUT_TO_TRY: [
+    'Would your code work for the following input?',
+    'How about the following input? Would your code still work?',
+    'Would your code work with the following?',
```
What happens when you run your code with this input? rather than this line.

rabidbit replied:
Revised as suggested.

seanlip commented on client/question/questionSpec.js:
```diff
@@ -92,4 +102,19 @@ describe('question.js', function() {
+      function() {
+        var correctnessFeedbackTextKeys =
+          Object.keys(CORRECTNESS_FEEDBACK_TEXT);
+        expect(correctnessFeedbackTextKeys).toContain(
```
Can you check for exact match to the list of feedback types, rather than just containment?

rabidbit replied:
toContain requires an exact match for strings. E.g., toContain would fail for just 'OUTPUT_'. The test checks to make sure the defined feedback type constant values are used in CORRECTNESS_FEEDBACK_TEXT.

Are you suggesting to do the reverse check as well? I.e., for each key in CORRECTNESS_FEEDBACK_TEXT, make sure there is some constant defined with the same value as the key?

seanlip replied:
Yes, that's right (re your second paragraph).

rabidbit replied:
Done

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +73,25 @@ tie.factory('FeedbackGeneratorService', [
     var consecutiveSameRuntimeErrorCounter = 0;
 
     /**
+     * Object to hold arrays indexes (values) of available correctness feedback
```
Nit: arrays --> array

rabidbit replied:
Fixed. Thanks

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -53,6 +73,25 @@ tie.factory('FeedbackGeneratorService', [
+     *
+     * @type {number}
+     */
+    var previousTestSuiteId = '';
```
Initialize to null? Weird to have a var that's supposed to be a string end up being a number.

Also something seems wrong here. We have test suite IDs, but those are strings, and they're IDs, not numerical indexes.

rabidbit replied:
I'm not sure I follow. Am I assigning numbers to this? Does var testSuiteId = testSuites[j].getId() return a number? I thought it returns a string.

seanlip replied:
Ignore first paragraph -- I got confused by the docstring. I think the docstring is wrong here in that it refers to a 'number' type and an index rather than ID.

rabidbit replied:
Ah! that should have been 'string'. Fixed.

seanlip replied:
Thanks! Just to make sure you caught it, note the point about index --> ID too. They mean different things: an index is a numeric index into an array, and an ID is a unique identifier.

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +101,51 @@ tie.factory('FeedbackGeneratorService', [
+     * @private
+     */
+    var _getCorrectnessFeedbackString = function(feedbackType) {
+      if (!availableCorrectnessFeedbackIndexes[feedbackType]) {
```
if (!availableCorrectnessFeedbackIndexes.hasOwnProperty(feedbackType)) {

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -62,6 +101,51 @@ tie.factory('FeedbackGeneratorService', [
+          Number.call, Number);
+      }
+      var randomArrayIndex =
+        _getRandomInt(
```
Nit: move this to end of previous line.

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +341,95 @@ tie.factory('FeedbackGeneratorService', [
+        previousTestSuiteId = testSuiteId;
+        // Catch regressions to errors previously encountered and passed.
+        // If certain feedback was given before, the user passes the test,
+        // then later regress, do not repeat feedback that was given before.
```
regress --> regresses

rabidbit replied:
Fixed

seanlip commented on client/question/services/FeedbackGeneratorService.js:
```diff
@@ -257,40 +341,95 @@ tie.factory('FeedbackGeneratorService', [
+      }
+      // If the suite ID corresponds to the sample input, the question will
+      // already have displayed the input and expected output, so we advance
+      // the correctness state to "expected output available".
```
"expected output displayed".

rabidbit replied:
Fixed. thanks

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -171,61 +167,131 @@ describe('FeedbackGeneratorService', function() {
-        allowedOutputs: ['a']
-      });
+    it('should present expected output second', function() {
+      var correctnessFeedbackParagraphs =
```
Don't need to assign to correctnessFeedbackParagraphs here, since you immediately overwrite it. I.e, here, just do `FeedbackGeneratorService...` and then in the next stanza do `var correctnessFeedbackParagraphs = ...`

Ditto below.

rabidbit replied:
Ah, fixed.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -171,61 +167,131 @@ describe('FeedbackGeneratorService', function() {
+          correctnessFeedbackParagraphs[0].getContent());
+        expect(correctnessFeedbackParagraphs[0].isTextParagraph()).toEqual(
+          true);
+        expect(correctnessFeedbackParagraphs[1].isOutputParagraph()).toEqual(
```
Verify the content of this paragraph? Ditto in other tests below.

rabidbit replied:
Done

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -1279,5 +1345,35 @@ describe('FeedbackGeneratorService', function() {
+        for (var i = 0; i < maxTries; i++) {
+          var randomNumber = FeedbackGeneratorService._getRandomInt(min, max);
+          expect(typeof randomNumber).toEqual('number');
+          expect(randomNumber).toBeGreaterThan(-1);
```
It would probably be a bit stronger if you had "greater than -0.00001" and "less than 3.00001".

Or, better: not.toBeLessThan(0) and similar. See https://stackoverflow.com/questions/24090270/how-can-i-test-that-a-value-is-greater-than-or-equal-to-in-jasmine

rabidbit replied:
Yeah, .not. is better, changed.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -1279,5 +1345,35 @@ describe('FeedbackGeneratorService', function() {
+      function() {
+        var selectedFeedbckText = '';
+        for (var correctnessFeedbackType in CORRECTNESS_FEEDBACK_TEXT) {
+          selectedFeedbckText =
```
"feedback" is misspelled in this variable name.

rabidbit replied:
Ah, sorry. thanks for the catch!

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -1279,5 +1345,35 @@ describe('FeedbackGeneratorService', function() {
+  describe('_getCorrectnessFeedbackString', function() {
+    it('should return a feedback string from CORRECTNESS_FEEDBACK_TEXT',
+      function() {
+        var selectedFeedbckText = '';
```
I don't think you need this declaration; just put a var two lines below.

rabidbit replied:
Fixed

seanlip commented on client/question/services/SolutionHandlerServiceSpec.js:
```diff
@@ -454,8 +456,18 @@ describe('SolutionHandlerService', function() {
-              'Your code produced the following result:');
+            var expectedFeedbackTextResult = 'valid feedback text';
+            var actualFeedbackTextResult = 'invalid feedback text';
+            var actualFeedbackText = feedback.getParagraphs()[0].getContent();
```
Don't do this -- the type should be deterministic, and it should be possible to just do something like `expect(CORRECTNESS_FEEDBACK_TEXT[...].includes(actualFeedbackTextResult)).toBe(true)`.

rabidbit replied:
The way it's set up, it needs to traverse multiple feedback type arrays. Modified it a bit, but please let me know if there is a better way.

seanlip replied:
I don't understand the need to traverse the multiple arrays. Shouldn't the expected typeKey be deterministic (for this particular input)? I can understand the uncertainty in the .includes() part, since we're randomly generating that, but I would expect the typeKey to be deterministic.

rabidbit replied:
OK, I think I changed it they way you are describing.

seanlip commented on client/question/services/SolutionHandlerServiceSpec.js:
```diff
@@ -560,8 +572,18 @@ describe('SolutionHandlerService', function() {
+          for (var typeKey in CORRECTNESS_FEEDBACK_TEXT) {
+            if (CORRECTNESS_FEEDBACK_TEXT[typeKey].includes(
+              actualFeedbackText)) {
+              actualFeedbackTextResult = expectedFeedbackTextResult;
```
Ditto -- don't have significant logic or vars that maintain state in tests.

rabidbit commented:
Think I addressed comments. PTAL.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -207,13 +207,20 @@ describe('FeedbackGeneratorService', function() {
           true);
-        expect(correctnessFeedbackParagraphs.length).toEqual(2);
+        var expectedOutputParagraph =
+          'Input: "' + sampleInputTestSuite.testCase.input +
```
Hardcode this fully, don't interpolate. This test is supposed to fail when changes to the code result in the output string not matching this one; we should have a "golden string" to compare against.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -222,76 +229,94 @@ describe('FeedbackGeneratorService', function() {
       expect(correctnessFeedbackParagraphs[1].isCodeParagraph()).toEqual(
         true);
-      expect(correctnessFeedbackParagraphs.length).toEqual(2);
+      var expectedInputCodeParagraph =
```
Hardcode fully.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -222,76 +229,94 @@ describe('FeedbackGeneratorService', function() {
       expect(correctnessFeedbackParagraphs[1].isCodeParagraph()).toEqual(
         true);
-      expect(correctnessFeedbackParagraphs.length).toEqual(2);
+      var expectedExpectedOutputParagraph =
```
Hardcode fully.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -222,76 +229,94 @@ describe('FeedbackGeneratorService', function() {
       expect(correctnessFeedbackParagraphs[1].isOutputParagraph()).toEqual(
         true);
-      expect(correctnessFeedbackParagraphs.length).toEqual(2);
+      var expectedOutputParagraph =
```
Hardcode fully.

seanlip commented on client/question/services/FeedbackGeneratorServiceSpec.js:
```diff
@@ -222,76 +229,94 @@ describe('FeedbackGeneratorService', function() {
       expect(correctnessFeedbackParagraphs[1].isCodeParagraph()).toEqual(
         true);
-      expect(correctnessFeedbackParagraphs.length).toEqual(2);
+      var expectedRegressionParagraph =
```
Hardcode fully.
